Fixes for ec2_ami tests

willthames opened the pull request:
##### SUMMARY

Several tests were marked as FIXME and should have been fixed with
the boto3 move. (I hadn't noticed this when merging ec2_ami)

Improved tags output. Add purge_tags option (default: no)

Allow description and tags update

Return launch_permissions

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
ec2_ami

##### ANSIBLE VERSION
```
ansible 2.5.0 (devel 310eb833a4) last updated 2017/10/30 10:21:33 (GMT +1000)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/will/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/will/src/ansible/lib/ansible
  executable location = /home/will/src/ansible/bin/ansible
  python version = 2.7.13 (default, Sep  5 2017, 08:53:59) [GCC 7.1.1 20170622 (Red Hat 7.1.1-3)]
```

willthames commented:
cc @wilvk

willthames committed:
Fixes for ec2_ami tests

Several tests were marked as FIXME and should have been fixed with
the boto3 move.

Improved tags output. Add purge_tags option (default: no)

Allow description and tags update

Return launch_permissions

wilvk commented:
@willthames wow! quick work! looks good!

gundalow commented:
`test/integration/targets/ec2_ami/tasks/main.yml:106` is failing with
```
{
    "assertion": "result.changed", 
    "changed": false, 
    "evaluated_to": false
}
```

Commit:
Allow empty launch permissions for image creation

Empty launch permissions should work the same way for image
creation as no launch permissions

willthames committed:
Cope with ephemeral devices in AMI block device mapping

Ephemeral devices can appear in AMI block devices, and
this information should be returned

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_ami.py:
```diff
@@ -542,69 +535,102 @@ def deregister_image(module, connection):
+
+    desired_tags = module.params.get('tags')
+    current_tags = boto3_tag_list_to_ansible_dict(image.get('Tags'))
+    tags_to_add, tags_to_remove = compare_aws_tags(current_tags, desired_tags, purge_tags=module.params.get('purge_tags'))
```
This should probably check if tags is not None first.
If the user doesn't specify tags in their play:
```
The full traceback is:
Traceback (most recent call last):
  File "/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_1M9gC7/ansible_module_ec2_ami.py", line 699, in <module>
    main()
  File "/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_1M9gC7/ansible_module_ec2_ami.py", line 692, in main
    update_image(module, connection, module.params.get('image_id'))
  File "/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_1M9gC7/ansible_module_ec2_ami.py", line 578, in update_image
    tags_to_add, tags_to_remove = compare_aws_tags(current_tags, desired_tags, purge_tags=module.params.get('purge_tags'))
  File "/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_1M9gC7/ansible_modlib.zip/ansible/module_utils/ec2.py", line 757, in compare_aws_tags
TypeError: argument of type 'NoneType' is not iterable

fatal: [localhost]: FAILED! => {
    "changed": false,
    "module_stderr": "Traceback (most recent call last):\n  File \"/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_1M9gC7/ansible_module_ec2_ami.py\", line 699, in <module>\n    main()\n  File \"/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_1M9gC7/ansible_module_ec2_ami.py\", line 692, in main\n    update_image(module, connection, module.params.get('image_id'))\n  File \"/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_1M9gC7/ansible_module_ec2_ami.py\", line 578, in update_image\n    tags_to_add, tags_to_remove = compare_aws_tags(current_tags, desired_tags, purge_tags=module.params.get('purge_tags'))\n  File \"/var/folders/by/k8_fbl593dlctgqmwq5wzl2c0000gn/T/ansible_1M9gC7/ansible_modlib.zip/ansible/module_utils/ec2.py\", line 757, in compare_aws_tags\nTypeError: argument of type 'NoneType' is not iterable\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE",
    "rc": 0
}
```

willthames replied:
That's a good spot - you also made me realise that if setting `purge_tags` and not setting `tags` at all, it could remove all existing tags.

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_ami.py:
```diff
@@ -542,69 +535,102 @@ def deregister_image(module, connection):
+        to_add = [dict(Group=group) for group in to_add_groups] + [dict(UserId=user_id) for user_id in to_add_users]
+        to_remove = [dict(Group=group) for group in to_remove_groups] + [dict(UserId=user_id) for user_id in to_remove_users]
+
+        if to_add or to_remove:
```
Not really sure what the desired behavior for this bit is. If I run a task with
```
        launch_permissions:
          group_names: ['all']
          user_ids: ['123456789012']
```
and a subsequent task removes the group_name 'all':
```
        launch_permissions:
          group_names: []
```
should that also remove the user_ids? That is what happens now.

willthames replied:
Yes, that would be what happens. If you pass launch permissions, you have to pass all the launch permissions. I honestly don't think many people modify launch permissions other than adding an AWS account. Could perhaps be better documented but that could be addressed later.

willthames replied:
I've updated the docs for launch permissions. I don't think the behaviour changes with this PR.

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_ami.py:
```diff
@@ -542,69 +535,102 @@ def deregister_image(module, connection):
+            module.fail_json_aws(e, msg="Error setting description for image %s" % image_id)
+
+    if changed:
+        module.exit_json(msg="AMI launch permissions updated.", changed=True,
```
Is this msg necessary for backwards compatibility? Seems a little misleading since launch permissions may not have been updated.

willthames replied:
I don't really like the `msg` return values, but removing them seems a step too far - I've made this one more generic though

Commit:
Address review feedback

ryansb commented on lib/ansible/modules/cloud/amazon/ec2_ami.py:
```diff
@@ -542,69 +536,103 @@ def deregister_image(module, connection):
+    if launch_permissions is not None:
+        current_permissions = image['LaunchPermissions']
+
+        current_users = set([permission['UserId'] for permission in current_permissions if 'UserId' in permission])
```
These comprehensions don't need `[]` around them - `set(str(user_id) for user_id in launch_permissions.get('user_ids', []))` because 2.6+ handles comprehensions in function calls without them.

ryansb reviewed (approved):
Minor tweak for unnecessary list permissions, I'll do it when squashing.

ryansb commented:
Merged as 60b29cf57d
